go/controller/internal/controller/clustersync_controller.go

```diff
@@ -0,0 +1,224 @@
+type ClusterSyncReconciler struct {
+	client.Client
+
+	ConsoleClient consoleclient.ConsoleClient
+	Scheme        *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=deployments.plural.sh,resources=clustersyncs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=deployments.plural.sh,resources=clustersyncs/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=deployments.plural.sh,resources=clustersyncs/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.16.3/pkg/reconcile
+func (r *ClusterSyncReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, retErr error) {
+	logger := log.FromContext(ctx)
+	logger.V(5).Info("reconciling ClusterSync")
+
+	clusterSync := new(v1alpha1.ClusterSync)
+	if err := r.Get(ctx, req.NamespacedName, clusterSync); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	utils.MarkCondition(clusterSync.SetCondition, v1alpha1.ReadyConditionType, v1.ConditionFalse, v1alpha1.ReadyConditionReason, "")
+	utils.MarkCondition(clusterSync.SetCondition, v1alpha1.SynchronizedConditionType, v1.ConditionFalse, v1alpha1.SynchronizedConditionReason, "")
+
+	scope, err := NewDefaultScope(ctx, r.Client, clusterSync)
+	if err != nil {
+		utils.MarkCondition(clusterSync.SetCondition, v1alpha1.SynchronizedConditionType, v1.ConditionFalse, v1alpha1.SynchronizedConditionReasonError, err.Error())
+		return ctrl.Result{}, err
+	}
+
+	// Always patch object when exiting this function, so we can persist any object changes.
+	defer func() {
+		if err := scope.PatchObject(); err != nil && retErr == nil {
+			retErr = err
+		}
+	}()
+
+	project, res, err := GetProject(ctx, r.Client, r.Scheme, clusterSync)
+	if res != nil || err != nil {
+		return handleRequeue(res, err, clusterSync.SetCondition)
+	}
+
+	yamlTemplateBytes, err := yaml.Marshal(clusterSync.Spec.ClusterSpec)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	yamlTemplate := string(yamlTemplateBytes)
+	var allErrs []error
+
+	pager := r.ListClusters(ctx, project.Status.ID, clusterSync.Spec.Tags)
+	for pager.HasNext() {
+		clusters, err := pager.NextPage()
+		if err != nil {
+			logger.Error(err, "failed to fetch cluster list")
+			return ctrl.Result{}, err
+		}
+
+		for _, cluster := range clusters {
+			clusterCRD, u, err := templateCluster(cluster, yamlTemplate)
+			if err != nil {
+				allErrs = append(allErrs, err)
+				logger.Error(err, "failed to template cluster")
+				continue
+			}
+			if clusterCRD.Namespace == "" {
+				clusterCRD.Namespace = clusterSync.Namespace
+			}
+			existingCluster := &v1alpha1.Cluster{}
+			if err := r.Client.Get(ctx, types.NamespacedName{Name: clusterCRD.Name, Namespace: clusterCRD.Namespace}, existingCluster); err != nil {
```

Comment:
also this is a pretty big function, i think we should refactor the actual crud of the cr into a separate function, and leave this to just handle pagination and filtering out dangling clusters to be pruned at the end.